File: go/0040M-组合总和II.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-13 10:18:03
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	ans := make([][]int, 0)
	vis := make([]int, 0)
	var dfs func(index, sum int)
	dfs = func(index, sum int) {
		if sum == target {
			tmp := make([]int, len(vis))
			copy(tmp, vis)
			ans = append(ans, tmp)
			return
		}
		for i := index; i < len(candidates) && sum+candidates[i] <= target; i++ {
			if i > index && candidates[i] == candidates[i-1] { // 去重同一层
				continue
			}
			vis = append(vis, candidates[i])
			dfs(i+1, sum+candidates[i])
			vis = vis[:len(vis)-1]
		}
	}
	sort.Ints(candidates)
	dfs(0, 0)
	return ans
}

File: cpp/0040M-组合总和II.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-26 14:43:44
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<vector<int>> ans;
    vector<int> tmp;
    bool vis[101] = {false};

    void dfs(vector<int>& candidates, int target, int idx, int sum) {
        if (sum == target) {
            ans.push_back(tmp);
            return;
        }
        for (int i = idx;
             i < candidates.size() && sum + candidates[i] <= target; i++) {
            if (i > 0 && candidates[i] == candidates[i - 1] && !vis[i - 1])
                continue;  // 去重
            vis[i] = true;
            tmp.push_back(candidates[i]);
            dfs(candidates, target, i + 1, sum + candidates[i]);
            tmp.pop_back();
            vis[i] = false;
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0, 0);
        return ans;
    }
};